op-node/rollup/derive/span_batch.go

```diff
@@ -44,80 +47,142 @@ type RawSpanBatch struct {
+	return nil
+}
+
+// decodeL1OriginCheck parses data into bp.decodeL1OriginCheck
+func (bp *spanBatchPrefix) decodeL1OriginCheck(r *bytes.Reader) error {
+	bp.l1OriginCheck = make([]byte, 20)
+	_, err := io.ReadFull(r, bp.l1OriginCheck)
 	if err != nil {
 		return fmt.Errorf("failed to read l1 origin check: %w", err)
 	}
 	return nil
 }
 
-// decodePayload parses data into b.spanBatchPayload
-func (b *RawSpanBatch) decodePayload(r *bytes.Reader) error {
+// decodePrefix parses data into bp.spanBatchPrefix
+func (bp *spanBatchPrefix) decodePrefix(r *bytes.Reader) error {
+	if err := bp.decodeRelTimestamp(r); err != nil {
+		return err
+	}
+	if err := bp.decodeL1OriginNum(r); err != nil {
+		return err
+	}
+	if err := bp.decodeParentCheck(r); err != nil {
+		return err
+	}
+	if err := bp.decodeL1OriginCheck(r); err != nil {
+		return err
+	}
+	return nil
+}
+
+// decodeBlockCount parses data into bp.blockCount
+func (bp *spanBatchPayload) decodeBlockCount(r *bytes.Reader) error {
 	blockCount, err := binary.ReadUvarint(r)
+	bp.blockCount = blockCount
 	// TODO: check block count is not too large
 	if err != nil {
 		return fmt.Errorf("failed to read block count: %w", err)
 	}
-	originBitBufferLen := blockCount / 8
-	if blockCount%8 != 0 {
-		originBitBufferLen++
-	}
-	originBitBuffer := make([]byte, originBitBufferLen)
-	_, err = io.ReadFull(r, originBitBuffer)
-	if err != nil {
-		return fmt.Errorf("failed to read origin bits: %w", err)
+	return nil
+}
+
+// decodeBlockTxCounts parses data into bp.blockTxCounts
+// and sets bp.txs.totalBlockTxCount as sum(bp.blockTxCounts)
+func (bp *spanBatchPayload) decodeBlockTxCounts(r *bytes.Reader) error {
+	if bp.txs == nil {
+		bp.txs = &spanBatchTxs{}
 	}
-	b.decodeOriginBits(originBitBuffer, blockCount)
-	blockTxCounts := make([]uint64, blockCount)
+	var blockTxCounts []uint64
 	totalBlockTxCount := uint64(0)
-	for i := 0; i < int(blockCount); i++ {
+	for i := 0; i < int(bp.blockCount); i++ {
 		blockTxCount, err := binary.ReadUvarint(r)
 		// TODO: check blockTxCount is not too large
 		if err != nil {
 			return fmt.Errorf("failed to read block tx count: %w", err)
 		}
-		blockTxCounts[i] = blockTxCount
+		blockTxCounts = append(blockTxCounts, blockTxCount)
 		totalBlockTxCount += blockTxCount
 	}
-	b.blockCount = blockCount
-	b.blockTxCounts = blockTxCounts
-	b.txs = &spanBatchTxs{}
-	b.txs.totalBlockTxCount = totalBlockTxCount
-	if err = b.txs.decode(r); err != nil {
+	bp.blockTxCounts = blockTxCounts
+	bp.txs.totalBlockTxCount = totalBlockTxCount
+	return nil
+}
+
+// decodePayload parses data into bp.spanBatchPayload
+func (bp *spanBatchPayload) decodePayload(r *bytes.Reader) error {
+	if bp.txs == nil {
+		bp.txs = &spanBatchTxs{}
+	}
+	if err := bp.decodeBlockCount(r); err != nil {
+		return err
+	}
+	if err := bp.decodeOriginBits(r); err != nil {
+		return err
+	}
+	if err := bp.decodeBlockTxCounts(r); err != nil {
+		return err
+	}
+	if err := bp.txs.decode(r); err != nil {
```

Comment:
We must set `totalBlockTxCount` before decoding `spanBatchTxs`. why don't you add sanity check for `totalBlockTxCount` in `txs.decode()`?

Reply:
Answered above.